### ISSUE TITLE
SHLVL for PowerShell - How many PowerShells deep am I?

### ISSUE THREAD

**Issue by rjmholt** [PowerShell/PowerShell#9613]
# Summary of the new feature/enhancement

Some POSIX shells have a variable called `SHLVL` that tracks how nested your shell session is. [This SO question](https://unix.stackexchange.com/questions/373704/how-many-shells-deep-i-am) goes into good depth on the topic.

I tend to find I need to know if I'm in a nested PowerShell session or not, particularly because:

- Developing binary modules, I often use a nested PS process to deal with type unloading
- Things like `$env:PSModulePath` can be twisted by nesting PowerShell sessions

I currently have a function in my profile to get information on this:

```powershell
function Get-PSParentProcesses
{
    $ps = [System.Collections.Stack]::new()
    $p = Get-Process -Id $pid
    do
    {
        if ($p.ProcessName -in "pwsh","pwsh-preview","powershell")
        {
            $ps.Push($p)
        }
        $p = $p.Parent
    }
    while ($p)

    return $ps
}
```

But given that `bash` already has functionality built-in to track this concept, it seemed worth opening a feature proposal for.

Some things to consider:

- Do we track other shells or other PowerShells? Environment variables aren't suited to this, but it's useful information
- Perhaps this variable is better off not as an environment variable but an automatic variable with real object information, or even as a cmdlet?

**Comment by vexx32** [comment]
This already exists, doesn't it? Check the default `prompt` function; it uses nesting information, right?

**Comment by fMichaleczek** [comment]
@vexx32 no, it doesn't exist.  the issue is about nested process when the prompt display nested pipeline/runspace/choosetheappropriateword.

**Comment by fMichaleczek** [comment]
I have the same issue, I will be happy to see that in the prompt by default. If not, I could set a custom prompt.

**Comment by iSazonov** [comment]
It seems rarely used feature. So I believe it is better implement it as external community module.

I would always first of all see if PowerShell could make smarter things than existing analogues before adding new code in PowerShell. I mean that we could think more about problems with nested shells (processes) and maybe find a smart and magic solution(s).

**Comment by bpayette** [comment]
I do find this extremely useful. As a result, I've been using an environment variable __ShellDepth since the beginning of PowerShell. It gets set in my profile 
```powershell
$ENV:__ShellDepth = [int] $ENV:__ShellDepth + 1
```
and echoed in my prompt. But, given how easy it is to implement externally and the fact that it probably isn't necessary for the average user, I'm not sure we'd want to implement in PowerShell itself.

**Comment by iSazonov** [comment]
> I do find this extremely useful. 

Not clear why. What is scenarios which need this? New PowerShell subprocess inherits nothing (except envs). I guess the work style comes from single windows console in Unix-s. I would suggest that most developers in the modern world prefer a multi-window interface which greatly improves developer productivity.

**Comment by rjmholt** [comment]
> Not clear why. What is scenarios which need this?

Personally I live on the keyboard and don't like using the mouse or interacting with GUIs (why I love terminal editors and why it's frustrating that Windows terminals don't play so well with them). I also tend to work and think in a stack-like way, pushing and popping contexts (and I feel like this is fairly common).

So when I need to get a clean PowerShell process to push a new context I don't want to do any more than type `pwsh<Enter>`. But unlike a computer I don't always remember how many stack frames deep I am.

GUIs cross-platform don't really have a convenient, fast, common, keyboard-native way to:
- Open a new PowerShell window from PowerShell
- Switch to that Window
- Execute things
- Close the Window
- Go back to the original PowerShell process

But even if they did, the fact that the GUI can do it has never really been a good argument that PowerShell shouldn't (quite the opposite I would say).

None of these are arguments this should be in PowerShell though! Me opening an issue is based on:
1. I have a need,
2. bash has a builtin concept to serve that need,
3. Do others have that need?
4. Is it worth emulating bash and making it a builtin, should it take a different form, should it be part of a module somewhere, or should it just be left up to individuals to implement themselves?

**Comment by iSazonov** [comment]
@rjmholt Thanks for great comment!

It should be in the issue description as motivation.
I like to improve single console experience too. (But I do not like primitive copy-paste features and want to see power and magic thing in PowerShell)

> >  I also tend to work and think in a stack-like way

Since PowerShell processes is independent what do you think about that we could generalize from stack-like "structure" to others? I mean for example that you could switch context by Tab (or Alt-Arrow) in circular or something else; get IntelliSense; get named contexts.

**Comment by bpayette** [comment]
@iSazonov A simple example for where this is useful: I'm a developer creating cmdlets. I don't want to load the cmdlet into the current shell because I can't unload it so I spawn a new shell and do my testing. Now after some time , i want to go back to my original shell but I can't remember if I've already done that or not. So having the shell depth in my prompt is very useful.

**Comment by chuanjiao10** [comment]
because I can't unload it--------remove-module.
I can't remember if I've already done that or not. ------open two ps windows on win or two terminal.
The variables common to the two powershell processes can only be environment variables.
For powershell, we should give higher priority to important enhancements,
eg: sftp drive ,
eg: give powershell release to eol Linux distribution one more year.

**Comment by SeeminglyScience** [comment]
> Now after some time , i want to go back to my original shell but I can't remember if I've already done that or not. So having the shell depth in my prompt is very useful.

I don't think a day goes by where I don't accidental type `exit` one time too many and sigh as I watch my perfectly set up window... disappear.  <sub>Obviously it's not that hard to open a new one, it's just *strangely* irritating.</sub>

+1

**Comment by rjmholt** [comment]
Very frustrating, especially in Windows when PS Core hangs on application cleanup for some reason

**Comment by DHowett** [comment]
Curiously, there's also `$NestedPromptLevel`, which is incremented by `$Host.EnterNestedPrompt()`. It's not a subshell, but it _is_ something you can exit from. It's a bit more useful combined with `Push-Location`/`Pop-Location`.

```
(libra) ~ % gc function:/enter-subshell

        Param (
                [string]$Path
        )
        Push-Location -StackName:"SubshellStack" -Path:$Path -ErrorAction:Stop > $Null
        $Host.EnterNestedPrompt()
        Pop-Location -StackName:"SubshellStack" > $Null

(libra) ~ % enter-subshell
(libra) ~ <1> % exit
(libra) ~ %
```

**Comment by iSazonov** [comment]
I see a conflict - how would we distinguish between nested prompt and nested subshell?

Another thought is about smart exit which prevents from accidental exit.

**Comment by SeeminglyScience** [comment]
> I see a conflict - how would we distinguish between nested prompt and nested subshell?

Should be pretty easy, nested prompts are controlled by the PSHost.  It's an entirely PowerShell concept. 

@bpayette's environment variable method would do the trick, you'd just have to put it in `UnmanagedPSEntry` or `ConsoleHost`.

**Comment by damianpowell** [comment]
My solution to this is fairly straight forward. I just add the line below to my `$profile.CurrentUserAllHosts` file. It's useful for Starship's SHLVL module.

```powershell
$env:SHLVL = [int]$env:SHLVL + 1
```

This approach plays nice whether the environment already has `SHLVL` defined, or not.

**Comment by dkaszews** [comment]
@damianpowell Perfect! It even interacts with bash as expected:

```shell
$ echo $SHLVL
1
$ pwsh
> $env:SHLVL  
2
> bash
$ echo $SHLVL
3
$ pwsh
> $env:SHLVL
4
```

**Comment by pfmoore** [comment]
@dkaszews The big problem with this approach (which I use currently) is that it goes wrong when you start a new terminal explicitly, `$SHLVL` doesn't get reset. This breaks the interpretation that I want, which is "`$SHLVL` is the number of times I need to type `exit` before the console closes".

To see this, just to `start pwsh` in a powershell window. In the new window, `$env:SHLVL` will be 2.

**Comment by pfmoore** [comment]
Here's a more robust approach. It checks the current console HWND, and uses that to determine whether to increment or restart `SHLVL`.

```
Add-Type -Namespace Win32 -Name Console -MemberDefinition @"
[DllImport("kernel32.dll")]
public static extern System.IntPtr GetConsoleWindow();
"@

$currentConsole = [Win32.Console]::GetConsoleWindow()

if ($env:POWERSHELL_CONSOLE_ID -eq "$currentConsole") {
    # Same console: increment
    $env:SHLVL = [int]$env:SHLVL + 1
}
else {
    # New console: reset
    $env:SHLVL = 1
}

$env:POWERSHELL_CONSOLE_ID = "$currentConsole"
```
